### PR TITLE
Add Data.Functor.Classes/Hashable1/NFData1 instances

### LINKED ISSUE #4: Oleg's to do list
- [x] `Strict` type-class #3
- [x] Add flagged dependency on `assoc` (instances for Pair and Either) #6 
- [x] Add changelog #5 
- [x] Add strict `These` #9 
- [x] `src/` directory
- [x] Safe Haskell #7 
- [ ] Eq1/Ord1/Show1/Read1/Hashable1/NFData1 instances #11

### PR DESCRIPTION
**Pull request by phadej**
Closes #4

### CODE REVIEW

File: strict/src/Data/Strict/Maybe.hs [hunk 1]
```diff
@@ -67,7 +81,17 @@ import           Data.Data           (Data (..), Typeable1 (..))
...
+#endif
+
+#if MIN_VERSION_deepseq(1,4,3)
+import Control.DeepSeq (NFData1 (..))
+#endif
+
+#ifdef LIFTED_FUNCTOR_CLASSES
+import Data.Functor.Classes (Eq1 (..), Ord1 (..), Read1 (..), Show1 (..))
+#else
+import Data.Functor.Classes (Eq1 (..), Ord1 (..), Read1 (..), Show1 (..))
```
Comment (infinity0) [inline strict/src/Data/Strict/Maybe.hs:94]: Both branches here are identical, or is there a reason to do this on purpose (future extensions)?

File: strict/src/Data/Strict/Either.hs [hunk 2]
```diff
@@ -6,6 +6,18 @@
 {-# LANGUAGE DeriveGeneric      #-}
 #endif
 
+#if MIN_VERSION_base(4,9,0)
+#define LIFTED_FUNCTOR_CLASSES 1
+#else
+#if MIN_VERSION_transformers(0,5,0)
+#define LIFTED_FUNCTOR_CLASSES 1
+#else
```
Comment (infinity0) [inline strict/src/Data/Strict/Either.hs:14]: You can shorten `#else #if` to `#elif` then we can have fewer `#endif` at the end.

(Same for the other files.)

File: strict/src/Data/Strict/Maybe.hs (hunk 1 above)
Comment (phadej) [reply strict/src/Data/Strict/Maybe.hs:94]: Yeah, I'd like to keep structure ~similar.

File: strict/src/Data/Strict/Either.hs (hunk 2 above)
Comment (phadej) [reply strict/src/Data/Strict/Either.hs:14]: We rather not. See https://github.com/isomorphism/these/pull/146

**Comment by phadej** [comment]
After this is merged, I'll move `strict-base-types` to this repository and adapt it. Let's see what's left there (besides `lens` instances).

File: strict/src/Data/Strict/Either.hs (hunk 2 above)
Comment (infinity0) [reply strict/src/Data/Strict/Either.hs:14]: This sucks, but OK. :(

Comment (phadej) [reply strict/src/Data/Strict/Either.hs:14]: I had the very same feeling when looking at the patch and reason with `these`.

CentOS-7:
- Full updates: Q4 2020
- Maintenance Update: June 30th, 2024

Just four more years!

**Pull request merged**